src/common/subprocess/subprocess.h

```diff
@@ -0,0 +1,370 @@
+#define _FLUX_CORE_SUBPROCESS_H
+
+#include <flux/core.h>
+
+/*
+ *  flux_cmd_t: An object that defines a command to be run, either
+ *   remotely or as a child of the current process. Includes cmdline
+ *   arguments, environment, and working directory. A flux_cmd_t is
+ *   used to create Flux subprocesses.
+ */
+typedef struct flux_command flux_cmd_t;
+
+/*
+ *  flux_subprocess_t: A subprocess is an instantiation of a command
+ *   as a remote or local process. A subprocess has a state (e.g.
+ *   initialized, starting, running, exited, completed), a PID, and
+ *   a rank if running remotely.
+ */
+typedef struct flux_subprocess flux_subprocess_t;
+
+/*  flux_subprocess_server_t: Handler for a subprocess remote server */
+typedef struct flux_subprocess_server flux_subprocess_server_t;
+
+/*
+ * Subprocess states, on changes, will lead to calls to
+ * on_state_change below.
+ *
+ * Possible state changes:
+ *
+ * init -> started
+ * started -> exec failed
+ * started -> running
+ * running -> exited
+ * any state -> failed
+ */
+typedef enum {
+    FLUX_SUBPROCESS_INIT,         /* initial state */
+    FLUX_SUBPROCESS_STARTED,      /* fork() has been issued/requested */
+    FLUX_SUBPROCESS_EXEC_FAILED,  /* exec(2) has failed, only for rexec() */
+    FLUX_SUBPROCESS_RUNNING,      /* exec(2) has been called */
+    FLUX_SUBPROCESS_EXITED,       /* process has exited */
+    FLUX_SUBPROCESS_FAILED,       /* internal failure, catch all for
+                                   * all other errors */
+} flux_subprocess_state_t;
+
+/*
+ * Subprocess flags
+ */
+enum {
+    /* flux_exec(): let parent stdin, stdout, stderr, carry to child.
+     * Do not create "STDIN", "STDOUT", or "STDERR" channels.  Subsequently,
+     * flux_subprocess_write()/close()/read()/read_line() will fail on
+     * streams of "STDIN", "STDOUT", or "STDERR".
+     */
+    FLUX_SUBPROCESS_FLAGS_STDIO_FALLTHROUGH = 1,
+    /* flux_exec(): call setpgrp() before exec(2) */
+    FLUX_SUBPROCESS_FLAGS_SETPGRP = 2,
+};
+
+/*
+ *  Typedefs for subprocess hooks and callbacks:
+ *
+ */
+typedef void (*flux_subprocess_f) (flux_subprocess_t *p);
+typedef void (*flux_subprocess_output_f) (flux_subprocess_t *p,
+                                          const char *stream);
+typedef void (*flux_subprocess_state_f) (flux_subprocess_t *p,
+                                         flux_subprocess_state_t state);
+
+/*
+ *  Functions for event-driven subprocess handling:
+ *
+ */
+typedef struct {
+    flux_subprocess_f on_completion;    /* Process exited and all I/O
+                                         * complete, will not be
+                                         * called if EXEC_FAILED or
+                                         * FAILED states reached.
+                                         */
+    flux_subprocess_state_f on_state_change;  /* Process state change          */
+    flux_subprocess_output_f on_channel_out; /* Read from channel when ready   */
```

Comment:
In the case of `on_channel_out` *and* `on_stdout` or `on_stderr`, do both callbacks get called, or do `on_stdout,on_stderr` take precendence?

I can't remember if we decided there was no use case for a `flux_subprocess_channel_fd`? As long as the PMI_FD case is handled, I think we've covered the main use case for now.

Reply:
Only ```on_stdout``` and ```on_stderr``` get called.  Effectively, the "stdout" and "stderr" channels/streams are treated as special cases and get their own callback.  I do see how this can be confusing.  For example, if you specify ```on_channel_out``` but NOT ```on_stdout```, would you expect stdout to go to ```on_channel_out```?  At the moment, the answer is no, it doesn't get sent there.

We could collapse all of these callbacks into just one ```on_channel_out```, and the user would have to ```!strcmp (stream, "STDOUT")``` for stdout?  What do you think?

Yeah, we decided to remove ```flux_subprocess_channel_fd()``` and all channel communication is via ```flux_subprocess_read()``` and ```flux_subprocess_write()``` calls now.

Reply:
Looking through use cases and thinking about it, I ***think*** collapsing it all into one output is a good idea.  Perhaps I won't know for sure until I implement it.  The most common case is to not do any channels, you only care about stdout/stderr.  In the one case we use channels (```flux-start```), we have the parent stdin/stdout/stderr fallthrough to the child process.  So we actually don't need to do the ```strcmp```.

Can you imagine other channel use cases in the future?